src/Build/BackEnd/Components/Scheduler/Scheduler.cs

```diff
@@ -176,8 +177,10 @@ internal class Scheduler : IScheduler
         /// </summary>
         public Scheduler()
         {
-            _debugDumpState = Environment.GetEnvironmentVariable("MSBUILDDEBUGSCHEDULER") == "1";
-            _debugDumpPath = Environment.GetEnvironmentVariable("MSBUILDDEBUGPATH");
+            _debugDumpState = Traits.Instance.DebugScheduler;
+            _debugDumpPath = Traits.Instance.DebugEngine
+                ? DebugUtils.DebugDumpPath()
```

Comment:
~~I'd put MSBUILDDEBUGPATH as priority if it's defined. That way, someone can overwrite the default from DebugDumpPath.~~

If DebugDumpPath defaults to Environment.GetEnvironmentVariable("MSBUILDDEBUGPATH") (as long as it has a value), why not just always take DebugUtils.DebugDumpPath? It could change behavior in a surprising way, but it makes more sense to me. (Maybe a change wave?)

Reply:
Good point, put it behind a change wave. I also replaced all the other direct reads of MSBuildDebugScheduler and MSBuildDebugPath with Traits and DebugUtils.DebugDumpPath